### DISCUSSION TITLE
Link and Dedupe

### DISCUSSION THREAD

**Discussion by hunsicke** [moj-analytical-services/splink#1601] [Q&A] [answered]
Are there any tutorials on splink using the link and dedupe setting?  

I am just guessing how these various settings work.  My guess is that dedupe is just linkage where the two sets being linked are the same.  Link and dedupe would be deduping where the set is the union of the two sets being linked, and where one of the linkage fields is set of origin.  

Is that right, or is there more to it?  Is it possible to define the overall expected probability of linkage separately from the overall expected probability of a duplicate in each set, so there are three different lambda values?

**Comment by ADBond** [comment, marked as answer]
That's more or less it - although in link+dedupe you don't necessarily _need_ to take into account the set of origin. I don't believe there are any specific tutorials on `link_and_dedupe` - there is just a little information on [the link type page](https://moj-analytical-services.github.io/splink/topic_guides/splink_fundamentals/link_type.html).

The difference between the three types is _purely_ in which records are (potentially) paired together during training and for prediction:
* `dedupe_only` accepts only a single table, and will consider all unique pairs between rows of this table
* `link_only` accepts two or more tables, and will consider all pairs of records from distinct tables, but not within-table pairings
* `link_and_dedupe` accepts two or more tables, and considers all unique pairs of records including between-table and within-table pairings. This is the same as doing a union of all the tables beforehand, and doing a `dedupe_only` on that union

Aside from that, the setting up of comparisons and parameters are the same for all types. There is only a single lambda value for the model for the overall prior, but if you wanted to account for differences should that be relevant for your use-case there are a couple of options:
* include a comparison in your model which takes into account the dataset of origin, so that the m/u values will account for any differences. I don't know how well this would work in practice.
* train separate `dedupe_only` models on your individual datasets, and then run a `link_only` model between them. Each of these models will have its own lambda, and you can vary the comparisons between them. This may be particularly useful if you have fields unique to either dataset which will be useful for deduping, or if you have many varied datasets, and is an approach that we sometimes use in practice.

**Reply by RobinL** [reply]
Spot on.  Just a couple more notes to expand on what Andy has said:

The decision to go with one `link_and_dedupe` vs three models (two `link_and_dedupe` and `link_only`) is basically just a complexity/accuracy tradeoff.  

If you believe the 'data generating process' for the two datasets is similar or the same, then a single model may suffice.   
If they're very different, then you probably want the three models.

By 'similar' I probably mean:
- probability of duplicates/links among random comparisons similar
- cardinality of columns similar
- data errors (typos etc) have similar prevalence in both datasts

For what it's worth, for the production models we run, we have separate `dedupe_only` models for each dataset and then a `link_only` one

**Reply by medwar99** [reply]
Hi both, thanks for the insight, this is really useful and the 3 model approach is the way I am going. Would it be possible to elaborate a little more on how you actually apply the subsequent `link_only`? 

I have two datasets where some specific rules are easier to implement with a `dedupe_only`, and these separate dedupes have provided nice pairwise predictions and clusters within the respective datasets. The aim here isn't really to remove duplicates in the dataset I suppose, but rather provide clusters of records which then want linking by the subsequent linker. 

Where would the `link_only` come in? I build a third linker and pass in the two datasets, but how could that take into account the previous dedupe clusters and predictions? Could I also instantiate the m/u parameters based on the two prior models? Should I be linking on the predicted outputs of the two dedupe linkers instead (i.e. cluster-to-cluster link)?
